# Unix Compression of hg19 Genomes
```
pi:ababaian
start: 2017 01 03
complete : 2017 01 03
```
## Introduction

It's quite a simple idea here. As information content goes up in a set, the compressability of that file decreases. Notably DNA sequences do not really compress very well using standard unix algorithms.

### Hypothesis

* The compression ratio of non-standard encodings of the genome will be higher then the standard encoding.

* The difference in compression ratio between the non-standard encodings will more closely relflect information differences.

Note: This isn't perfect since any single alt-encoding doesn't contain all the information of the standard encoding but it will still give a measure IF this type of compression can work at all.

## Objective

- For each encdoing, extract chr10
- Apply each compression algorithm available in unix to chr10_enc.fa
- Measure the size of the output in bits

## Materials and Methods
Everything should be capable on most unix systems

In [4]:
# Global Variables
SERRATUS='/home/artem/Serratus/'

In [3]:
uname -svrpo
lshw -class cpu

Linux 4.4.0-57-generic #78-Ubuntu SMP Fri Dec 9 23:50:32 UTC 2016 x86_64 GNU/Linux
  *-cpu
       product: Intel(R) Core(TM) i7-2620M CPU @ 2.70GHz
       vendor: Intel Corp.
       physical id: 1
       bus info: cpu@0
       size: 1299MHz
       capacity: 3400MHz
       width: 64 bits
       capabilities: fpu fpu_exception wp vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp x86-64 constant_tsc arch_perfmon pebs bts nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx lahf_lm epb tpr_shadow vnmi flexpriority ept vpid xsaveopt dtherm ida arat pln pts cpufreq


In [6]:
# Navigate to hg19 resource folder
cd $SERRATUS/resources/hg19

# Fasta Index each genome
  fastaindex hg19.fa hg19.index
  fastaindex hg19_ry.fa hg19_ry.index
  fastaindex hg19_sw.fa hg19_sw.index
  fastaindex hg19_mk.fa hg19_mk.index
  fastaindex hg19_h.fa hg19_h.index

** FATAL ERROR **: Index path [hg19.index] already exists
exiting ...


In [23]:
# Extract chromosome 10 from each file
cd $SERRATUS/resources/hg19
  fastafetch hg19.fa hg19.index chr10 > chr10.fa
  fastafetch hg19_ry.fa hg19_ry.index chr10 > chr10_ry.fa
  fastafetch hg19_sw.fa hg19_sw.index ghr10 > chr10_sw.fa
  fastafetch hg19_mk.fa hg19_mk.index ahr10 > chr10_mk.fa
  fastafetch hg19_h.fa hg19_h.index chr10 > chr10_h.fa
  
# Move to chr10 folder
    mkdir -p chr10
    mv chr10*.fa ./chr10/
    cd chr10

In [12]:
# File sizes
ls -l

# Chromsoome 10 files
CHRTEN=$(ls)

total 671264
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10.fa
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_h.fa
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_mk.fa
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_ry.fa
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_sw.fa


In [13]:
# Bzip2
bzip2 --version

for FA in $CHRTEN
do
    bzip2 -dzc $FA > $FA.bzip2
done

# ls -l

bzip2, a block-sorting file compressor.  Version 1.0.6, 6-Sept-2010.
   
   Copyright (C) 1996-2010 by Julian Seward.
   
   This program is free software; you can redistribute it and/or modify
   it under the terms set out in the LICENSE file, which is included
   in the bzip2-1.0.6 source distribution.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   LICENSE file for more details.
   


In [14]:
ls -l

total 796456
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10.fa
-rw-rw-r-- 1 artem artem  37024762 Jan  3 12:57 chr10.fa.bzip2
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_h.fa
-rw-rw-r-- 1 artem artem  17208582 Jan  3 12:57 chr10_h.fa.bzip2
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_mk.fa
-rw-rw-r-- 1 artem artem  23576941 Jan  3 12:58 chr10_mk.fa.bzip2
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_ry.fa
-rw-rw-r-- 1 artem artem  22630493 Jan  3 12:58 chr10_ry.fa.bzip2
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_sw.fa
-rw-rw-r-- 1 artem artem  27740515 Jan  3 12:58 chr10_sw.fa.bzip2


In [17]:
# gzip
gzip --version

for FA in $CHRTEN
do
    gzip -c $FA > $FA.gzip
done

ls -l

gzip 1.6
Copyright (C) 2007, 2010, 2011 Free Software Foundation, Inc.
Copyright (C) 1993 Jean-loup Gailly.
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.

Written by Jean-loup Gailly.
total 815052
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10.fa
-rw-rw-r-- 1 artem artem  42312950 Jan  3 13:46 chr10.fa.gzip
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_h.fa
-rw-rw-r-- 1 artem artem  21541861 Jan  3 13:46 chr10_h.fa.gzip
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_mk.fa
-rw-rw-r-- 1 artem artem  25970824 Jan  3 13:47 chr10_mk.fa.gzip
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_ry.fa
-rw-rw-r-- 1 artem artem  25590474 Jan  3 13:47 chr10_ry.fa.gzip
-rw-rw-r-- 1 artem artem 137470965 Jan  3 12:50 chr10_sw.fa
-rw-rw-r-- 1 artem artem  31812467 Jan  3 13:47 chr10_sw.fa.gzip


In [24]:
# Remove Ns
# Change all cases to uppercase
sed 's/N//g' chr10.fa | sed '/^$/d' - | tr A-Z a-z > chr10.tmp
  mv chr10.tmp chr10.fa
        
sed 's/N//g' chr10_ry.fa | sed '/^$/d' - | tr A-Z a-z > chr10.tmp
  mv chr10.tmp chr10_ry.fa
sed 's/N//g' chr10_sw.fa | sed '/^$/d' - | tr A-Z a-z > chr10.tmp
  mv chr10.tmp chr10_sw.fa
sed 's/N//g' chr10_mk.fa | sed '/^$/d' - | tr A-Z a-z > chr10.tmp
  mv chr10.tmp chr10_mk.fa
sed 's/N//g' chr10_h.fa  | sed '/^$/d' - | tr A-Z a-z > chr10.tmp
  mv chr10.tmp chr10_h.fa

In [25]:
xz -V

xz (XZ Utils) 5.1.0alpha
liblzma 5.1.0alpha


In [26]:
# bzip2 & gzip & xz

for FA in $CHRTEN
do
    xz -c $FA > $FA.xz
    bzip2 -dzc $FA > $FA.bzip2
    gzip -c $FA > $FA.gzip
done

ls -l

total 995744
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10.fa
-rw-rw-r-- 1 artem artem  36475036 Jan  3 14:02 chr10.fa.bzip2
-rw-rw-r-- 1 artem artem  39559214 Jan  3 14:03 chr10.fa.gzip
-rw-rw-r-- 1 artem artem  32561152 Jan  3 14:02 chr10.fa.xz
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10_h.fa
-rw-rw-r-- 1 artem artem  16617129 Jan  3 14:08 chr10_h.fa.bzip2
-rw-rw-r-- 1 artem artem  20231126 Jan  3 14:09 chr10_h.fa.gzip
-rw-rw-r-- 1 artem artem  14816516 Jan  3 14:08 chr10_h.fa.xz
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10_mk.fa
-rw-rw-r-- 1 artem artem  22982657 Jan  3 14:15 chr10_mk.fa.bzip2
-rw-rw-r-- 1 artem artem  24013369 Jan  3 14:15 chr10_mk.fa.gzip
-rw-rw-r-- 1 artem artem  18682512 Jan  3 14:15 chr10_mk.fa.xz
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10_ry.fa
-rw-rw-r-- 1 artem artem  22080751 Jan  3 14:22 chr10_ry.fa.bzip2
-rw-rw-r-- 1 artem artem  23786189 Jan  3 14:22 chr10_ry.fa.gzip
-rw-rw-r-- 1 artem artem  17769008 Jan  3 14:21 ch

## Results & Discussion

At fundamental level the compression does appear to be working more efficiently in the the non-standard encodings over the standard encoding

|File|bits|Compression_Ratio|
|----|----|-----------------|
|chr10.fa|133190688|1|
|chr10.fa.bzip2|36475036|3.65155741|
|chr10.fa.gzip|39559214|3.366868917|
|chr10.fa.xz|32561152|4.090478371|
|chr10_h.fa|133190688|1|
|chr10_h.fa.bzip2|16617129|8.015264731|
|chr10_h.fa.gzip|20231126|6.58345403|
|chr10_h.fa.xz|14816516|8.989339194|
|chr10_mk.fa|133190688|1|
|chr10_mk.fa.bzip2|22982657|5.795269363|
|chr10_mk.fa.gzip|24013369|5.546522356|
|chr10_mk.fa.xz|18682512|7.129163787|
|chr10_ry.fa|133190688|1|
|chr10_ry.fa.bzip2|22080751|6.031981793|
|chr10_ry.fa.gzip|23786189|5.59949675|
|chr10_ry.fa.xz|17769008|7.495673816|
|chr10_sw.fa|133190688|1|
|chr10_sw.fa.bzip2|22161976|6.00987421|
|chr10_sw.fa.gzip|23702722|5.619214873|
|chr10_sw.fa.xz|18187388|7.323244437|

None of the alternative encodings ry / sw / mk compressed more then twice as much as chr10. I think this an issue with the encoding since these are ASCII files, every character is 8 bytes regardless. The higher compression ratios in non-standard encoding suggests a higher information density but this isn't complete.


What I need to do is covert chr10.fa / chr10_xy.fa into true binary files where the ATCG is 2-bit encoded and SW/RY/MK is 1-bit encoded instead of ASCII bytes. Then compress these files respectively and compare compression ratios.


## Recoding as binary

To encode the genome files as a simple binary, further manipulate the .fa files to reduce them to simple binary.


In [29]:
# Two-bit encoding for ATCG encoding
    sed 1d chr10.fa |\
    sed 's/a/10/g' - |\
    sed 's/t/00/g' - |\
    sed 's/g/11/g' - |\
    sed 's/c/01/g' - |\
    tr -d '[\n]' | perl -lpe '$_=pack"B*",$_' > chr10.bit

In [30]:
# Binary encoding for Non-standard encodings
# RY 
    sed 1d chr10_ry.fa |\
    sed 's/a/1/g' - |\
    sed 's/c/0/g' - |\
    tr -d '[\n]' | perl -lpe '$_=pack"B*",$_' > chr10_ry.bit
    
# SW
    sed 1d chr10_sw.fa |\
    sed 's/t/1/g' - |\
    sed 's/g/0/g' - |\
    tr -d '[\n]' | perl -lpe '$_=pack"B*",$_' > chr10_sw.bit
    
# MK
    sed 1d chr10_mk.fa |\
    sed 's/a/1/g' - |\
    sed 's/g/0/g' - |\
    tr -d '[\n]' | perl -lpe '$_=pack"B*",$_' > chr10_mk.bit
    
# H 
    sed 1d chr10_h.fa |\
    sed 's/g/1/g' - |\
    sed 's/c/0/g' - |\
    tr -d '[\n]' | perl -lpe '$_=pack"B*",$_' > chr10_h.bit

In [31]:
CHRBIT=$(ls *.bit)

# bzip2 & gzip & xz

for FA in $CHRBIT
do
    7z a $FA.7z $FA
    xz -c $FA > $FA.xz
    bzip2 -dzc $FA > $FA.bzip2
    gzip -c $FA > $FA.gzip
done

ls -l

total 1357048
-rw-rw-r-- 1 artem artem  32828686 Jan  3 17:20 chr10.bit
-rw-rw-r-- 1 artem artem  31232508 Jan  3 17:27 chr10.bit.bzip2
-rw-rw-r-- 1 artem artem  31302823 Jan  3 17:27 chr10.bit.gzip
-rw-rw-r-- 1 artem artem  29373000 Jan  3 17:27 chr10.bit.xz
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10.fa
-rw-rw-r-- 1 artem artem  36475036 Jan  3 14:02 chr10.fa.bzip2
-rw-rw-r-- 1 artem artem  39559214 Jan  3 14:03 chr10.fa.gzip
-rw-rw-r-- 1 artem artem  32561152 Jan  3 14:02 chr10.fa.xz
-rw-rw-r-- 1 artem artem  16414344 Jan  3 17:26 chr10_h.bit
-rw-rw-r-- 1 artem artem  13275813 Jan  3 17:27 chr10_h.bit.bzip2
-rw-rw-r-- 1 artem artem  12684403 Jan  3 17:27 chr10_h.bit.gzip
-rw-rw-r-- 1 artem artem  11511956 Jan  3 17:27 chr10_h.bit.xz
-rw-rw-r-- 1 artem artem 133190688 Jan  3 13:58 chr10_h.fa
-rw-rw-r-- 1 artem artem  16617129 Jan  3 14:08 chr10_h.fa.bzip2
-rw-rw-r-- 1 artem artem  20231126 Jan  3 14:09 chr10_h.fa.gzip
-rw-rw-r-- 1 artem artem  14816516 Jan  3 14:08 chr10_h.

In [2]:
cd /home/artem/Serratus/resources/hg19/chr10

CHRBIT=$(ls *.bit)

# add 7zip algorithm

for FA in $CHRBIT
do
    7z a $FA.7z $FA
done

ls -l


7-Zip [64] 9.20  Copyright (c) 1999-2010 Igor Pavlov  2010-11-18
p7zip Version 9.20 (locale=en_CA.UTF-8,Utf16=on,HugeFiles=on,4 CPUs)

Scanning

Updating archive chr10.bit.7z

Compressing  chr10.bit          0%    0%    0%    0%    0%    0%    1%    1%    1%    1%    1%    2%    2%    2%    2%    2%    3%    3%    3%    3%    3%    4%    4%    4%    4%    4%    5%    5%    5%    5%    5%    6%    6%    6%    6%    6%    7%    7%    7%    7%    7%    8%    8%    8%    8%    8%    9%    9%    9%    9%    9%   10%   10%   10%   10%   10%   11%   11%   11%   11%   11%   12%   12%   12%   12%   12%

## Results & Discussion II: IO

|File|8 * bits|Compression Ratio|bits/base|
|----|--------|-----------------|---------|
|chr10.bit|32828686|1|2.00000003|
|chr10.bit.7z|29300481|1.120414576|1.785053562|
|chr10.bit.bzip2|31232508|1.051106302|1.902757148|
|chr10.bit.gzip|31302823|1.04874522|1.907040902|
|chr10.bit.xz|29373000|1.117648385|1.789471589|
|chr10_h.bit|16414344|1|1.000000076|
|chr10_h.bit.7z|11502774|1.426990046|0.70077579|
|chr10_h.bit.bzip2|13275813|1.236409702|0.808793456|
|chr10_h.bit.gzip|12684403|1.294057276|0.772763381|
|chr10_h.bit.xz|11511956|1.425851871|0.701335178|
|chr10_mk.bit|16414344|1|1.000000076|
|chr10_mk.bit.7z|15875166|1.033963613|0.967152096|
|chr10_mk.bit.bzip2|16318994|1.005842885|0.994191132|
|chr10_mk.bit.gzip|16314822|1.006100097|0.993936964|
|chr10_mk.bit.xz|15886272|1.033240775|0.9678287|
|chr10_ry.bit|16414344|1|1.000000076|
|chr10_ry.bit.7z|14788686|1.109925791|0.900961204|
|chr10_ry.bit.bzip2|15639416|1.049549676|0.952789657|
|chr10_ry.bit.gzip|15753561|1.041944993|0.959743636|
|chr10_ry.bit.xz|14805456|1.108668588|0.901982871|
|chr10_sw.bit|16414344|1|1.000000076|
|chr10_sw.bit.7z|15357556|1.068812251|0.935618089|
|chr10_sw.bit.bzip2|16086943|1.020351971|0.980054045|
|chr10_sw.bit.gzip|15886961|1.033195965|0.967870675|
|chr10_sw.bit.xz|15368488|1.068051978|0.936284092|


This is more in line with what I was expecting, a much lower compression ratio after converting the data to a raw binary format. Using these algorithms, all non-standard encodings performed worse then the standard encoding with respect to compression ratio. The H encoding, which is lossy performed better, but this can also be because there is lost information in it (i.e. it has less then half the data of the standard encoding).

It is also noteworthy that using this method, RY encoding seems to compress the best (most information content) compared to MK or SW.
